cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -324,6 +335,37 @@ private Put convertFieldsToBytes(Collection<Field<?>> fields) throws InvalidFiel
     return put;
   }
 
+  /**
+   * Updates fields in the row and converts to a {@link Put} to write to table. The primary key must be provided.
+   *
+   * @param fields the fields to update
+   * @return a PUT object
+   * @throws InvalidFieldException if primary keys are missing or the column is not in schema
+   */
+  private Put updateFieldsToBytes(Collection<Field<?>> fields) throws InvalidFieldException {
+    Set<String> fieldNames = fields.stream().map(Field::getName).collect(Collectors.toSet());
+    if (!fieldNames.containsAll(schema.getPrimaryKeys())) {
+      throw new InvalidFieldException(schema.getTableId(), fields,
+                                      String.format("Given fields %s does not contain all the " +
+                                                      "primary keys %s", fieldNames, schema.getPrimaryKeys()));
+    }
+    Collection<Field<?>> keyFields = fields.stream()
+      .filter(field -> schema.isPrimaryKeyColumn(field.getName())).collect(Collectors.toList());
+    byte[] primaryKey = convertKeyToBytes(keyFields, false);
+    Put put = new Put(primaryKey);
+    Row row = table.get(primaryKey);
+    for (Map.Entry<byte[], byte[]> entry : row.getColumns().entrySet()) {
+      String columnName = Bytes.toString(entry.getKey());
+      Optional<Field<?>> field = fields.stream().filter(f -> f.getName().equals(columnName)).findFirst();
```

Comment:
It's better to get the fields in a `Set`, then loop the fields (and remove field from the Set), then loop the missing fields from the Set.

Reply:
used a Map<String, Field> instead because we need to lookup column name.